File: src/common/xPlaylistLib/XplPLSPlaylist.m
//
//  XplPlsPlaylist.m
//  Xapl
//
//  Created by Julia on 5/30/09.
//  Copyright 2009 __MyCompanyName__. All rights reserved.
//

#import "XplPLSPlaylist.h"
#import "XplData.h"

@implementation XplPLSPlaylist

+ (BOOL) readData:(NSMutableArray *) data fromFile:(NSString *)file
{
	int numberOfEntries = -1, currentEntry = 1;
	BOOL playlistFlag = NO, isVersion2 = NO;
	NSString *fileData, *currentLine, *lowerLine, *comp, *dataFinder, *filePath, *title, *length;
	NSArray *dataLines;
	NSEnumerator *enumerator;
	NSError *error = nil;
	XplData *plsData;
	
	// Clear contents of data array
	[data removeAllObjects];
	
	// Attempt to read file
	fileData = [NSString stringWithContentsOfFile:file encoding:NSASCIIStringEncoding error:&error];
	if (fileData == nil)
	{
		NSLog([NSString stringWithFormat:@"%@", error]);
		return NO;
	}
	
	// Normalize line endings to UNIX style
	fileData = [fileData stringByReplacingOccurrencesOfString:@"\r\n" withString:@"\n"];
	
	// Split file into lines
	dataLines = [fileData componentsSeparatedByString:@"\n"];
	
	enumerator = [dataLines objectEnumerator];	
	while (currentLine = (NSString *)[enumerator nextObject]) 
	{
		// Blank line
		if ( [currentLine length] == 0 ) continue;
		
		// Trim trailing and leading spaces
		currentLine = [currentLine stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
		
		// Blank line after trimming
		if ( [currentLine length] == 0 ) continue;
		
		// Lower case copy
		lowerLine = [currentLine lowercaseString];
		
		// Header information
		if ( !playlistFlag && [lowerLine compare:@"[playlist]"] == 0 )
		{
			playlistFlag = true;
			continue;
		}
		else if ( numberOfEntries < 0 && [[lowerLine substringToIndex:15] compare:@"numberofentries"] == NSOrderedSame )
		{
			numberOfEntries = [[lowerLine substringFromIndex:16] integerValue];
			continue;
		}
		// Playlist items
		else if ( playlistFlag && currentEntry <= numberOfEntries )
		{
			// File path
			if ( [filePath length] == 0 )
			{
				// Is the numbering correct?
				comp = [lowerLine substringWithRange:NSMakeRange(4, [lowerLine rangeOfString:@"="].location - 4)];
				if ( [comp compare:[NSString stringWithFormat:@"%u", currentEntry]] != NSOrderedSame )
					continue;
				
				// Store file path
				dataFinder = [NSString stringWithFormat:@"file%u=", currentEntry];				
				if ( [lowerLine rangeOfString:file].location != NSNotFound )
					filePath = [currentLine substringFromIndex:[dataFinder length]];
			}
			// Track title, optional
			else if ( [title length] == 0 )
			{
				// Is the numbering correct?
				comp = [lowerLine substringWithRange:NSMakeRange(5, [lowerLine rangeOfString:@"="].location - 5)];
				if ( [comp compare:[NSString stringWithFormat:@"%u", currentEntry]] != NSOrderedSame )
					continue;
				
				// Store track title
				dataFinder = [NSString stringWithFormat:@"title%u=", currentEntry];				
				if ( [lowerLine rangeOfString:file].location != NSNotFound )
					title = [currentLine substringFromIndex:[dataFinder length]];
			}
			// Track length
			else if ( [length length] == 0 )
			{
				// Is the numbering correct?
				comp = [lowerLine substringWithRange:NSMakeRange(6, [lowerLine rangeOfString:@"="].location - 6)];
				if ( [comp compare:[NSString stringWithFormat:@"%u", currentEntry]] != NSOrderedSame )
					continue;
				
				// Store track title
				dataFinder = [NSString stringWithFormat:@"length%u=", currentEntry];				
				if ( [lowerLine rangeOfString:file].location != NSNotFound )
					length = [currentLine substringFromIndex:[dataFinder length]];
			}
			
			// Have all required components
			if ( [filePath length] != 0 && [length length] != 0 && [[lowerLine substringToIndex:5] compare:@"title"] != NSOrderedSame )
			{
				plsData = [[XplData alloc] initWithPath:filePath];
				[plsData setLength:length];
				if ([title length] != 0)
					[plsData setTitle:title];
				
				[data addObject:plsData];
				filePath  = @"";
				title = @"";
				length = @"";
			}
		}
		// Footer
		else if ( playlistFlag && currentEntry > numberOfEntries )
		{
			if ( [lowerLine compare:@"version=2"] == NSOrderedSame )
			{
				isVersion2 = YES;
				break;
			}
		}
	}
	
	return YES;
}

+ (BOOL) writeData:(NSArray *)data toFile:(NSString *)file
{
	NSEnumerator *enumerator;
	NSMutableString *output = [[[NSMutableString alloc] init] autorelease];
	NSError *error = nil;
	XplData *item;
	int counter = 1;
	id object;
	
	// Header
	[output appendString:@"[Playlist]\n"];
	[output appendFormat:@"NumberOfEntries=%u\n", [data count]];
	
	// Write data
	enumerator = [data objectEnumerator];
	while (object = [enumerator nextObject])
	{
		if (![object isKindOfClass:[XplData class]])
			return NO;
		
		item = (XplData *)object;
		
		// File path
		[output appendFormat:@"File%u=%@\n", counter, [item path]];
		
		// Title (if applicable)
		if ([[item title] length] != 0)
			[output appendFormat:@"Title%u=%@\n", counter, [item title]];
		
		// Length (required, -1 inserted if none specified)
		if ([[item length] length] != 0)
			[output appendFormat:@"Length%u=%@\n", counter, [item length]];
		else
			[output appendFormat:@"Length%u=-1\n", counter];
				 
		 ++counter;
	}
	
	// Footer
	[output appendString:@"Version=2\n"];
	
	// Write out the file
	[output writeToFile:file atomically:YES encoding:NSUnicodeStringEncoding error:&error];
	
	if (error != nil)
		return NO;
	
	return YES;
}

+ (BOOL) writeBlankPlaylistToFile:(NSString *)file
{
	NSError *error = nil;
	[@"[Playlist]\nNumberOfEntries=0\nVersion=2" writeToFile:file atomically:YES encoding:NSUnicodeStringEncoding error:&error];
	
	if (error != nil)
		return NO;
	
	return YES;
}

+ (BOOL) appendData:(NSArray *)data toFile:(NSString *)file
{
	NSMutableArray *currentData = [[NSMutableArray alloc] init];
	NSEnumerator *enumerator;
	XplData *plsData;
	id object;
	
	// Load any current data from the file
	if ( ![XplPLSPlaylist readData:currentData fromFile:file] )
		return NO;
	
	// Append data items
	enumerator = [data objectEnumerator];
	while (object = [enumerator nextObject])
	{
		if (![object isKindOfClass:[XplData class]])
			return NO;
		
		plsData = (XplData *)object;
		[currentData addObject:plsData];
	}
	
	// Write out combined data
	return [XplPLSPlaylist writeData:currentData toFile:file];
}

+ (NSString *) formatExtension
{
	return @"pls";
}

+ (NSString *) fullFormatName
{
	return @"Generic Playlist File";
}

@end

File: src/common/xPlaylistLib/XplPLSPlaylist.cpp
#include "XplPLSPlaylist.h"
#include <sstream>
#include <fstream>

#ifdef _DEBUG
#include <crtdbg.h>
#define DEBUG_NEW new(_NORMAL_BLOCK ,__FILE__, __LINE__)
#else
#define DEBUG_NEW new
#endif
#ifdef _DEBUG
#define new DEBUG_NEW
#endif 

XplPLSPlaylist::XplPLSPlaylist( void ) : m_loaded( false )
{
	m_plData = new vector<XplData *>;
}

XplPLSPlaylist::XplPLSPlaylist( const string &plFile ) : m_plFile( string( plFile ) ), m_loaded( false )
{
	m_plData = new vector<XplData *>;
}
	
bool XplPLSPlaylist::ReadPlaylist( void )
{
	if ( m_plFile.empty() ) return false;
	return ReadPlaylist( m_plFile );
}

bool XplPLSPlaylist::ReadPlaylist( const string &file )
{
	// Clear old data
	FreeMemory();
	
	// Parse playlist file and load data into the plData array
	ifstream input( file.c_str() );
	if ( !input.is_open() )
		return false;

	long numberOfEntries = -1, currentEntry = 1;
	bool playlistFlag = false, isVersion2 = false;
	string line, lowerLine, filePath, title, length;
	stringstream buffer;
	while ( getline( input, line ) )
	{
		// Trim trailing and leading spaces
		Trim( &line );

		// Convert to lower case
		lowerLine = line;
		Lower( &lowerLine );

		// Blank line
		if ( lowerLine.empty() ) continue;

		// Comment line
		if ( line[0] == ';' ) continue;

		// Header information
		if ( !playlistFlag && lowerLine.compare("[playlist]") == 0 )
		{
			playlistFlag = true;
			continue;
		}
		else if ( numberOfEntries < 0 && lowerLine.substr(0, 15).compare("numberofentries") == 0 )
		{
			numberOfEntries = atoi(lowerLine.substr(16).c_str());
			continue;
		}
		// Playlist items
		else if ( playlistFlag && currentEntry <= numberOfEntries )
		{
			// File path
			if ( filePath.length() == 0 && lowerLine.substr(0, 4).compare("file") == 0 )
			{
				// Is the numbering correct?
				buffer.str("");
				buffer << currentEntry;
				string test = lowerLine.substr(4, lowerLine.find("=") - 4 );
				if ( lowerLine.substr(4, lowerLine.find("=") - 4 ).compare( buffer.str() ) != 0 )
					continue;
				
				// Store file path
				buffer.str("");
				buffer << "file" << currentEntry << "=";

				if ( lowerLine.find( buffer.str() ) != string::npos )
					filePath = line.substr( buffer.str().length() );
			}
			// Track title, optional
			if ( title.length() == 0  && lowerLine.substr(0, 5).compare("title") == 0 )
			{
				// Is the numbering correct?
				buffer.str("");
				buffer << currentEntry;
				if ( lowerLine.substr(5, lowerLine.find("=") - 5 ).compare( buffer.str() ) != 0 )
					continue;

				// Store track title
				buffer.str("");
				buffer << "title" << currentEntry << "=";

				if ( lowerLine.find( buffer.str() ) != string::npos )
					title = line.substr( buffer.str().length() );
			}
			// Track length
			if ( length.length() == 0 && lowerLine.substr(0, 6).compare("length") == 0 )
			{
				// Is the numbering correct?
				buffer.str("");
				buffer << currentEntry;
				if ( lowerLine.substr(6, lowerLine.find("=") - 6 ).compare( buffer.str() ) != 0 )
					continue;

				// Store the track length
				buffer.str("");
				buffer << "length" << currentEntry << "=";

				if ( lowerLine.find( buffer.str() ) != string::npos )
					length = line.substr( buffer.str().length() );
			}
			
			// Have all required components
			if ( filePath.length() != 0 && length.length() != 0 && lowerLine.substr( 0, 5 ).compare("title") != 0 )
			{
				XplData *data = new XplData( filePath, title, length );
				m_plData->push_back( data );
				++currentEntry;
				filePath.clear();
				title.clear();
				length.clear();
			}
		}
		// Footer
		else if ( playlistFlag && currentEntry > numberOfEntries )
		{
			if ( lowerLine.compare("version=2") == 0 )
			{
				isVersion2 = true;
				break;
			}
		}
	}

	input.close();
	return true;
}

bool XplPLSPlaylist::WritePlaylist( void )
{
	if ( m_plFile.empty() ) return false;
	return WritePlaylist( m_plFile );
}

bool XplPLSPlaylist::WritePlaylist( const string &file )
{
	ofstream plFile( file.c_str(), std::ios::trunc );
	if ( plFile.is_open() )
	{
		// Header
		stringstream header;
		header << "[Playlist]" << endl << "NumberOfEntries=" << m_plData->size() << endl;
		plFile.write( header.str().c_str(), header.str().length() );

		for ( size_t i = 0; i < m_plData->size(); ++i )
		{
			stringstream itemData;

			// File path
			itemData << "File" << (i + 1) << "=" << m_plData->at( i )->GetPath() << endl;
			
			// Title (if applicable)
			if ( m_plData->at( i )->GetTitle().length() > 0 )
				itemData << "Title" << (i + 1) << "=" << m_plData->at( i )->GetTitle() << endl;

			// Length (required, -1 inserted if none specified)
			if ( m_plData->at( i )->GetLength().length() > 0 )
				itemData << "Length" << (i + 1) << "=" << m_plData->at( i )->GetLength() << endl;
			else
				itemData << "Length" << (i + 1) << "=-1" << endl;

			// Write it out
			plFile.write( itemData.str().c_str(), itemData.str().length() );
		}

		// Footer
		stringstream footer;
		footer << endl << "Version=2";
		plFile.write( footer.str().c_str(), footer.str().length() );

		// Close file
		plFile.close();
		return true;
	}

	return false;
}

// Create a blank playlist
bool XplPLSPlaylist::NewPlaylist( const string &file )
{
	ofstream plFile( file.c_str(), std::ios::out );
	if ( plFile.is_open() )
	{
		// Header
		stringstream header;
		header << "[Playlist]" << endl << "NumberOfEntries=0" << endl;
		plFile.write( header.str().c_str(), header.str().length() );

		// Footer
		stringstream footer;
		footer << endl << "Version=2";
		plFile.write( footer.str().c_str(), footer.str().length() );

		plFile.close();
		return true;
	}

	return false;
}

bool XplPLSPlaylist::AppendToPlaylist( const string &file, XplData **data, size_t n )
{
	try
	{
		XplPLSPlaylist pl;
		pl.ReadPlaylist( file );

		for ( size_t i = 0; i < n; ++i )
			pl.m_plData->push_back( (XplData *) data[i]);

		pl.WritePlaylist( file );
		pl.FreeMemory();
	}
	catch (...)
	{
		return false;
	}

	return true;
}

void XplPLSPlaylist::FreeMemory( void )
{
	for ( vector<XplData *>::iterator it = m_plData->begin(); it != m_plData->end(); it++ )
		delete *it;
	
	m_plData->clear();
	m_loaded = false;
}

void XplPLSPlaylist::SetData( XplData **data, size_t n )
{
	if ( n == 0 ) return;

	FreeMemory();
	XplData *d;
	for ( size_t i = 0; i < n; ++i )
	{
		d = static_cast<XplData *>(data[i]);
		m_plData->push_back(d);
	}
}

void XplPLSPlaylist::GetData( long index, const string &key, string *data )
{
	if ( !m_loaded ) *data = "";
	
	if ( key.compare( "" ) == 0 || key.compare( "file" ) == 0 )
		data->assign( m_plData->at( index )->GetPath() );
	else if ( key.compare( "length" ) == 0 )
		*data = m_plData->at( index )->GetLength();
	else if ( key.compare( "title" ) == 0 )
		*data = m_plData->at( index )->GetTitle();
	else
		*data = "";
}

void XplPLSPlaylist::GetRawData( XplData **data )
{
	if ( data == NULL ) return;
	for ( size_t i = 0; i < m_plData->size(); ++i )
	{
		data[ i ] = new XplData( m_plData->at( i )->GetPath(),
									m_plData->at( i )->GetTitle(),
									m_plData->at( i )->GetLength() );
	}
}

bool XplPLSPlaylist::GetNext( string *data, const string &key )
{
	return false;
}

size_t XplPLSPlaylist::GetCount( void )
{
	return m_plData->size();
}

bool XplPLSPlaylist::KeyIsValid( const string &key )
{
	return key.compare( "file" ) == 0 ||
		key.compare( "length" ) == 0 ||
		key.compare( "title" ) == 0;
}

void XplPLSPlaylist::GetFormatExtension( string *ext ) const
{
	if ( ext != NULL )
		*ext = "pls";
}

void XplPLSPlaylist::GetFormatFullName( string *fullName ) const
{
	if ( fullName != NULL )
		*fullName = "Generic Playlist File";
}
